### PR TITLE
rabbitmq-ha add prometheus operator stats and alerts

### PR DESCRIPTION
**Pull request by bradenwright**
**What this PR does / why we need it**:
This PR allows for stats to be exposed and scraped by the prometheus operator.  I've also setup some default alerts that I think would be helpful.

And the ability to manually create a secret before running, to ensure secret doesn't change and doesn't have to be set via value every run.

**Special notes for your reviewer**:
@etiennetremel please review

### CODE REVIEW

**Comment by bradenwright** [comment]
@etiennetremel please review when you have a chance.

File: stable/rabbitmq-ha/templates/NOTES.txt
Comment (etiennetremel) [inline stable/rabbitmq-ha/templates/NOTES.txt]: I might be looking at something old but where does this come from, I only the following in the values.yaml:
```
secret:
  autogenerate: true
```

**Comment by etiennetremel** [comment]
Hi @bradenwright, thanks for spending the time to improve this chart.
I'm totally fine with the Prometheus exporter but assuming the user is using the Prometheus Operator seems wrong. Users should be able to spin-up their own version of Prometheus 1.x, 2.x or operator.
But maybe one of the repo admin can guide of us with best practices.

File: stable/rabbitmq-ha/values.yaml
Comment (etiennetremel) [inline stable/rabbitmq-ha/values.yaml]: This doesn't seem to be used anywhere?

File: stable/rabbitmq-ha/values.yaml
Comment (etiennetremel) [inline stable/rabbitmq-ha/values.yaml]: Could keep the file consistent and add descriptions for fields that require it?
Also, the README.md should be updated if you add new fields.

File: stable/rabbitmq-ha/values.yaml
Comment (etiennetremel) [inline stable/rabbitmq-ha/values.yaml]: I suggest removing the alerts section and flatten a bit as suggested in the chart best practices:
```
prometheusExporter:
  enabled: true
  [...]
```

**Review by etiennetremel — CHANGES_REQUESTED** [review]

**Comment by bradenwright** [comment]
I'll see about making those changes.

And I agree that others may very well want support for Prometheus without the Operator.  I don't know the details of implementing without the operator, if someone else wanted to contribute supporting it without operator I'd be for it.

File: stable/rabbitmq-ha/Chart.yaml
Comment (timstoop) [inline stable/rabbitmq-ha/Chart.yaml]: Your adding a new feature, so please bump the minor version.

**Review by timstoop — CHANGES_REQUESTED** [review]
With regards to Prometheus without the Operator, it works with simple Pod or Service Annotations that Prometheus can discover and use to scrape for a certain endpoint. So by simply allowing custom annotation to the exporter Pod you'd be handling that case as well, if you'd like.

**Comment by unguiculus** [comment]
The service monitor for the operator should be optional in case a users wants to configure Prometheus integration with annotation as @timstoop mentioned.

**Comment by bradenwright** [comment]
Ok so I was working on Kafka chart to add support for metrics as well.  I tried to keep changes here consistent with variable naming on Kafka chart:

https://github.com/kubernetes/charts/pull/5120

There was also an example of using Prometheus without the Operator, so I added code for that as well, but have no way to test it.

My apologies it took me a while to reply, other work trumped this for a little while.

**Comment by bradenwright** [comment]
@etiennetremel if you can look over PR again when you have a chance, it would be much appreciated

**Comment by etiennetremel** [comment]
Thanks, do you mind updating the README.md with the variable change you just made?

**Comment by bradenwright** [comment]
@etiennetremel I added to the readme.  I also tried to make some tweaks to secrets, looks like some stuff was merge since I opened my PR, so tried to find a middle ground for var naming since I was adding more granularity.  I'm open to suggestions.

**Comment by bradenwright** [comment]
@timstoop @etiennetremel when you get a chance

**Comment by bradenwright** [comment]
@timstoop @etiennetremel ping

**Review by etiennetremel — APPROVED** [review]
Thanks @bradenwright!

**Comment by bradenwright** [comment]
Thanks for reviewing!  Sorry for the gap in between I know you basically did it 2x

**Comment by bradenwright** [comment]
Fixed merge conflict on chart version so @timstoop @etiennetremel @unguiculus if you can do your thing again.

**Review by unguiculus — CHANGES_REQUESTED** [review]
I don't think you need both `overrideName` and `autogenerate`. I'd prefer you keep `existingSecret`. This is what's commonly used. Also, I think `autogenerate` is misleading. It implies that certs are generated which is not the case.

You could create a partial for the secret name. Here's an example: https://github.com/kubernetes/charts/blob/master/stable/mongodb-replicaset/templates/_helpers.tpl#L34-L54.

**Comment by bradenwright** [comment]
@unguiculus the use case I have that we implemented that way for is that we like to setup 1 secret per namespace in some situations and then have a few rabbitmq-ha releases using the same secret.  This wasn't possible with `existingSecret` approach.  

Do you want me to change the verbage of `autogenerate` to `existingSecret`? I don't mind that at all I'd just like to keep ability to override the name.  I don't see any drawback to having it, just gives more flexibility.

**Comment by bradenwright** [comment]
Certs have their own section so I did the identical thing there, have the ability to autogenerate and overrideName, I'd be open to moving them both to a more generic area, or renaming top level secret so its more descriptive if you have a better suggestion

**Comment by bradenwright** [comment]
@unguiculus I updated the variable naming to use `existing` instead of `autogenerate` as requested.  I left the ability to override the name of the secret b/c we use that feature as I mentioned above.  Hopefully this satisfies both of us

**Comment by bradenwright** [comment]
@unguiculus ping

**Comment by unguiculus** [comment]
There's a separate PR for an exporter. It's probably better to add it as a dependency once it's merged.

#5311

cc @gianrubio

**Comment by bradenwright** [comment]
@unguiculus @gianarb just to be clear this PR already includes the rabbitmq exporter

**Comment by bradenwright** [comment]
@unguiculus @gianarb I didn't do a separate chart b/c I am running the rabbitmq-exporter as a sidecar and there isn't a way to do that with a dependency.  Running as a sidecar feel appropriate to me as the monitoring scales automatically with the pods, especially for monitoring things like rabbitmq nodes that are up.  Kafka chart is running a jmx exporter as a sidecar fwiw.

**Comment by gianrubio** [comment]
I'd recommend to add the sidecar container in [rabbit-mq chart](https://github.com/kubernetes/charts/blob/master/stable/rabbitmq/README.md).

Adding a [`serviceMonitor`](https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/examples/basic-auth/service-monitor.yaml) makes sense in this case but not deploying prometheus-operator because it should be deployed in a separate chart.

**Comment by bradenwright** [comment]
@gianrubio not sure I fully follow, if I missed something please clarify.  

This chart has no dependencies on stable/rabbitmq chart, tbh I don't even know why both exist.  Really seems like the 2 should be merged together, as this chart can be run with 1 replica if desired, I really don't see the point of stable/rabbitmq, but that maybe another discussion.  If someone wants to add changes to stable/rabbitmq chart they can, I have no use case for stable/rabbitmq chart.  So feel that should be a separate PR.

This PR does deploy a `ServiceMonitor` but doesn't deploy Prometheus or Prometheus Operator, as you mention.

File: stable/rabbitmq-ha/templates/alerts.yaml
Comment (gianrubio) [inline stable/rabbitmq-ha/templates/alerts.yaml]: this line is duplicated on 41

**Comment by bradenwright** [comment]
Also I'm looking to fix merge conflicts and it seems like there was just a PR merged that used `customSecret: false` instead of `existingSecret` which was already being used for Certificate. 
 @unguiculus I still haven't got confirmation from you, its says changes requested but I've replied to that.   I'm going to change it to be `existingSecret` since you mentioned earlier on this PR wanting to be consistent.

If you can please confirm if other changes are needed

**Comment by bradenwright** [comment]
I updated the wording of variable with `customSecret`, `customConfigMap` to `existing` b/c I was asked to use wording for secret above, and then figured configmap should probably be consistent name with other variables.

**Comment by bradenwright** [comment]
@unguiculus @timstoop @gianrubio any comments, changes, etc.  I've tried to make the changes requested but haven't had any feedback in a min.

**Comment by unguiculus** [comment]
I still don't see why you need to override the name of the secret if it is created by the chart. Either you use an existing secret or you have the chart create it. Why would you want to change the name then?

Maybe I wasn't clear in my comments. I wanted you to change it such that `existingSecret` is not a boolean but refers to the name of the existing secret. If set, use it, and otherwise create a secret with the name being based on the `fullname` template.

**Comment by bradenwright** [comment]
@unguiculus I'm doing both using an existing secret and overriding the name so it can be used for a few rabbitmq clusters.  The secret defaults to using a name with the release in it so I need to do both.

**Comment by bradenwright** [comment]
I can make change you asked about having `existingSecret` double up as a boolean and override name.  The reason I didn't do that is I felt leaving it defaulting to template fullname seemed like an appropriate default, and to do that I'm gonna have to do some slight of hand in templates/_helpers, I thought it felt cleaner to do it in the values.

@unguiculus let me know if want me to make changes or not after explanation.

**Comment by steven-sheehy** [comment]
@bradenwright I added fullnameOverride override value which is used in template "rabbitmq-ha.fullname". The secret will use that fullnameOverride so you can set that value if you don't want secret name to contain release name. Will that cover your use case?

**Comment by bradenwright** [comment]
@steven-sheehy thanks for the suggestion, it wouldn't b/c I'd have to set the fullnameOverride to be the same for multiple helm installs, and they will conflict for resources chart tries to create.

**Comment by steven-sheehy** [comment]
Maybe due how long this PR has been opened and since there's disagreement on only the secret changes those can be separated into a different PR? I'm very interested in having the prometheus exporter support that you so kindly implemented be merged in. Just my observation as an outsider, feel free to disregard.

**Comment by bradenwright** [comment]
@steven-sheehy I hear you, I really want to get this merged too.  I've pushed back a deploy with my company 2x now, hoping to get this merged rather than deploying a fork.

I'll make changes that @unguiculus wants, I just need feedback at this point.  I feel its really close so I'd rather make the tweak and merge than go through another PR this late in the game.

**Comment by unguiculus** [comment]
I don't think it makes sense to install the chart multiple times when one installation creates the secret and the others use that secret making them dependent on this installation. Uninstalling the release that installed the secret would break all other. If you need a shared secret you should install it separately.

**Comment by bradenwright** [comment]
@unguiculus that's exactly what I'm trying to do (install secret separately), but after I manually create 1 secret I want to use it for multiple installs (I don't want to create a new secret for every install in the same namespace).  For instance we have a number of rabbitmq sandboxes and want to use the same existing secret for all of them.  My solution solves this issue.

**Comment by unguiculus** [comment]
So, I think we're on the same line, but there seems to be some misunderstanding. If `existingSecret` is not a flag but contains the name of the secret, why would you need a name override? Here's an example:

https://github.com/kubernetes/charts/blob/master/incubator/keycloak/templates/_helpers.tpl#L42-L51

There are several charts that do it like this.

**Comment by bradenwright** [comment]
@unguiculus I made the change requested.  I think it should be good to go now, but if anything else is needed let me know.

File: stable/rabbitmq-ha/values.yaml
Comment (unguiculus) [inline stable/rabbitmq-ha/values.yaml]: Use a stable version

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by bradenwright** [comment]
@unguiculus sorry that took a min, I was out of town and didn't have access to a computer.

**Review by unguiculus — APPROVED** [review]

**Comment by francesco-beccaria** [comment]
Hi @bradenwright, @unguiculus,
I have some doubts on how to use the changes made in this PR.
I installed `prometheus` using the stable chart, so I thought that meant I should not get the `operator` enabled, so I set:
```
operator:
  enabled: false
```
When I tried to upgrade my existing rabbitmq-ha release, Helm complained with the message:
```
Error: UPGRADE FAILED: failed to create resource: namespaces "monitoring" not found
```
I didn't modify the default `# Namespace Prometheus is installed in`
```
operator:
  serviceMonitor:
    namespace: monitoring
```
as it's indented under the `operator` section, which I disabled, and I was expecting it would not be considered.

I'm probably just confused.

Any chance you could help me understand?

Also, I noticed that the default for `prometheus.exporter.enabled` in the `Readme` is set to `true`, but I think it should be `false`? I know it's "just" the `Readme` but I thought I'd point it out.

Thanks in advance,
Fran

**Comment by francesco-beccaria** [comment]
To add to my confusion, I saw the `monitoring` `namespace` is also in the `alerts.yaml` file, but in this case it's hardcoded.
Is it supposed to be the same as in the `servicemonitor.yaml` file (`{{ .Values.prometheus.operator.serviceMonitor.namespace }}`) or is it a different thing?

**Pull request merged**
